### PR TITLE
Xenomorph Traps

### PR DESCRIPTION
**Pull request by jjpark-kb**
## About The Pull Request
Adds the ability for any Xenomorph to place traps within hugger holes.
The kinds of traps are: Acid gas, Neuro gas, Blind sting, Brute sting, Toxic sting, Wall Trap, R. Wall Trap, and Snare.
It takes 5 seconds to fill the trap, and costs 50 plasma.

## Why It's Good For The Game
Adds more flexibility in terms of play style. Marines need to be more careful of traps and be more aware of the environment that they are in.

## Changelog
:cl:
add: Added traps for hugger holes
/:cl:

### CODE REVIEW

**Comment by AtlasRules** [comment]
what is this? a DEFILER buff? 
will it finally not suck?
probably will but eh

**Comment by Hughgent** [comment]
Sorry to say but this isn't how you should perform this function. The reason being is that it doesn't cost any Xeno anything to fill a trap. yes it costs a Carrier to create a trap, but it should also cost the filling xeno some amount of plasma.

**Comment by jjpark-kb** [comment]
> Sorry to say but this isn't how you should perform this function. The reason being is that it doesn't cost any Xeno anything to fill a trap. yes it costs a Carrier to create a trap, but it should also cost the filling xeno some amount of plasma.

Oh yea, for sure. This is currently a draft. I am trying to get some cost on it and perhaps some restrictions on which castes can fill.

**Comment by Hughgent** [comment]
Discussed on discord a bit.
Needs to make it cost plasma, of some amount, probably 50.
Have the trap scale based on xeno tier.
Possibly include 'special' trap options based on caste.

File: code/game/objects/structures/xeno.dm [hunk 1]
```diff
@@ -115,6 +115,10 @@
 	var/obj/item/clothing/mask/facehugger/hugger = null
 	var/mob/living/linked_carrier //The carrier that placed us.
 
+	var/gastrap = null
+	var/tgastier = null
```
Comment (psykzz) [inline code/game/objects/structures/xeno.dm:119]: What does this stand for?

File: code/game/objects/structures/xeno.dm [hunk 2]
```diff
@@ -169,6 +191,35 @@
 	visible_message("<span class='warning'>[hugger] gets out of [src]!</span>")
 	hugger = null
 
+/obj/effect/alien/resin/trap/proc/acid_activate()
+	if(gastrap)
```
Comment (psykzz) [inline code/game/objects/structures/xeno.dm:195]: Dont really need this if, as its handled via the switch

File: code/game/objects/structures/xeno.dm (hunk 1 above)
Comment (jjpark-kb) [reply code/game/objects/structures/xeno.dm:119]: ah, I added the t to the gastier to not get weird shit. The gastrap itself is for what is chosen from the list that each xeno is given.

File: code/game/objects/structures/xeno.dm (hunk 2 above)
Comment (jjpark-kb) [reply code/game/objects/structures/xeno.dm:195]: alrighty, Ill remove the if(gastrap)

**Comment by jroinc** [comment]
Sleep gas is a bad idea.

What do the wall traps do?

**Comment by jjpark-kb** [comment]
> Sleep gas is a bad idea.
> 
> What do the wall traps do?

will remove sleep gas.
Wall traps spawn a resin wall if a marine walks on it.
Rwall traps are just thick resin walls.

**Comment by spookydonut** [comment]
legcuffing marines is a terrible idea imo, hugger traps are already loathed.

**Comment by jjpark-kb** [comment]
> legcuffing marines is a terrible idea imo, hugger traps are already loathed.

done, removed the snare portion.

File: code/modules/mob/living/carbon/xenomorph/xeno_defines.dm [hunk 3]
```diff
@@ -232,3 +232,7 @@
 	var/fire_luminosity = 0 //Luminosity of the current fire while burning
 
 	var/butchery_progress = 0
+
+	//Trap vars
+	var/trapchoices = list()
+	var/gastier = 5
```
Comment (psykzz) [inline code/modules/mob/living/carbon/xenomorph/xeno_defines.dm:238]: should the default really be a 5 here? wouldn't 1 be more appropriate so people don't get confused when adding a new caste?

Comment (psykzz) [inline code/modules/mob/living/carbon/xenomorph/xeno_defines.dm:238]: Also would you mind adding docs to these defines?
Something like this but with detail on what its used for

```suggestion
	/// different trap choices
	var/trapchoices = list()
	/// the level of gas trap
	var/gastier = 5
```

Comment (jjpark-kb) [reply code/modules/mob/living/carbon/xenomorph/xeno_defines.dm:238]: ah yes, Ill also do this. Sorry about that!

Comment (jjpark-kb) [reply code/modules/mob/living/carbon/xenomorph/xeno_defines.dm:238]: Ah, will fix it, sorry!

**Comment by MacBlaze1** [comment]
Could you uh add a check inbetween line 63 and 64 that goes like
var/list/potential_gasTrap = range(X,src)
for(gasTrap in potential_gasTrap)
  if(isType(gasTrap, /obj/effect/alien/resin/trap))
     if(gasTrap.gastrap !=null)
         to_chat(usr, "There are too many gasTraps nearby!")
         return

to make sure it cant be abused within the gas radius.

**Comment by spookydonut** [comment]
From the extensive testing this has had, I think this has a lot of gameplay problems and should be reworked before being test merged again.

**Comment by jjpark-kb** [comment]
max gas tier is 3, cooldown of 10 seconds for placing the traps, makes them not place-able next to each other.

**Comment by jroinc** [comment]
Make them non-shootable if you're nerfing them this much, with the non-stacking and limited range.

Also, make the single-target low-damage ones harder to see (the stings), but keep the others (Gasses, walls, and blinds) just as visible as they are now.

**Comment by jjpark-kb** [comment]
> Make them non-shootable if you're nerfing them this much, with the non-stacking and limited range.
> 
> Also, make the single-target low-damage ones harder to see (the stings), but keep the others (Gasses, walls, and blinds) just as visible as they are now.

I am pretty sure they are non-shootable? Not entirely sure. Ill check it out.
Ill also do that for the sting traps.

**Comment by jjpark-kb** [comment]
Im getting no updates and there isnt much I can do. Im closing this because I dont really want to deal with it anymore.

**Pull request closed without merging**